### PR TITLE
[feature] (vectorization)parquet push down support

### PR DESCRIPTION
**Pull request by wsjz**
# Proposed changes

Issue Number: close #xxx

## Problem Summary:

Describe the overview of changes.

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by wsjz** [comment]
这是来自QQ邮箱的假期自动回复邮件。
好，我最近正在休假中

 
您，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。

File: be/src/exec/arrow/parquet_row_group_reader.cpp [hunk 1]
```diff
@@ -0,0 +1,539 @@
...
+            if (conjunct_slot_id == tuple_desc->slots()[i]->id()) {
+                // Get conjuncts by conjunct_slot_id
+                auto iter = _slot_conjuncts.find(conjunct_slot_id);
+                if (_slot_conjuncts.end() == iter) {
+                    std::vector<ExprContext*> conjuncts;
+                    conjuncts.emplace_back(_conjunct_ctxs[conj_idx]);
+                    _slot_conjuncts.emplace(std::make_pair(conjunct_slot_id, conjuncts));
+                } else {
+                    std::vector<ExprContext*> conjuncts = iter->second;
+                    conjuncts.emplace_back(_conjunct_ctxs[conj_idx]);
```
Comment (yinzhijian) [inline be/src/exec/arrow/parquet_row_group_reader.cpp:172]: Are the conjuncts here useless?

File: be/src/exec/arrow/parquet_row_group_reader.cpp [hunk 2]
```diff
@@ -0,0 +1,539 @@
...
+                }
+            }
+        }
+    }
+}
+
+void RowGroupReader::_determine_filter_row_group(const std::vector<ExprContext*>& conjuncts,
```
Comment (yinzhijian) [inline be/src/exec/arrow/parquet_row_group_reader.cpp:179]: Why not return bool directly here, but use the parameter to return the result

File: be/src/exec/arrow/parquet_row_group_reader.cpp [hunk 3]
```diff
@@ -0,0 +1,539 @@
...
+        Expr* conjunct = conjuncts[i]->root();
+        if (TExprNodeType::BINARY_PRED == conjunct->node_type()) {
+            _eval_binary_predicate(conjuncts[i], min_bytes, max_bytes, need_filter);
+        } else if (TExprNodeType::IN_PRED == conjunct->node_type()) {
+            _eval_in_predicate(conjuncts[i], min_bytes, max_bytes, need_filter);
+        }
+    }
+}
+
+void RowGroupReader::_eval_binary_predicate(ExprContext* ctx, const char* min_bytes,
```
Comment (yinzhijian) [inline be/src/exec/arrow/parquet_row_group_reader.cpp:195]:
```suggestion
bool RowGroupReader::_eval_binary_predicate(ExprContext* ctx, const char* min_bytes,
```

File: be/src/exec/arrow/parquet_row_group_reader.cpp [hunk 4]
```diff
@@ -0,0 +1,539 @@
...
+    case TYPE_DATETIME: {
+        std::vector<char*> in_values;
+        for (auto val : in_pred_values) {
+            char* value = reinterpret_cast<char**>(val)[0];
+            in_values.emplace_back(value);
+        }
+        if (in_values.empty()) {
+            return false;
+        }
+        if (in_values.size() >= 2) {
```
Comment (yinzhijian) [inline be/src/exec/arrow/parquet_row_group_reader.cpp:279]: There seems to be no need to determine whether >= 2, both can use this logic

File: be/src/exec/arrow/parquet_row_group_reader.cpp [hunk 5]
```diff
@@ -0,0 +1,539 @@
...
+    return false;
+}
+
+bool RowGroupReader::_eval_gt(PrimitiveType conjunct_type, void* value, const char* max_bytes) {
+    switch (conjunct_type) {
+    case TYPE_TINYINT: {
+        _PLAIN_DECODE_SINGLE(int8_t, value, max_bytes, conjunct_value, max)
+        _FILTER_GROUP_BY_GT_PRED(conjunct_value, max)
+        break;
+    } break;
```
Comment (yinzhijian) [inline be/src/exec/arrow/parquet_row_group_reader.cpp:354]:
```suggestion
    }
```

**Comment by yinzhijian** [comment]
We may need to add a check to the unit tests to see if the filter function really works

File: be/src/exec/arrow/parquet_row_group_reader.cpp (hunk 1 above)
Comment (wsjz) [reply be/src/exec/arrow/parquet_row_group_reader.cpp:172]: it has been in map yet

File: be/src/exec/arrow/parquet_row_group_reader.cpp [hunk 6]
```diff
@@ -0,0 +1,567 @@
...
+            }
+            auto statistic = row_group_meta->ColumnChunk(parquet_col_id)->statistics();
+            if (!statistic->HasMinMax()) {
+                continue;
+            }
+            // Min-max of statistic is plain-encoded value
+            const std::string& min = statistic->EncodeMin();
+            const std::string& max = statistic->EncodeMax();
+
+            need_filter = _determine_filter_row_group(slot_iter->second, min, max);
```
Comment (SaintBacchus) [inline be/src/exec/arrow/parquet_row_group_reader.cpp:133]: is this `need_filter` determined by the last row_group?

**Comment by wsjz** [comment]
TPCH sql  completed
in be.conf  enable： parquet_predicate_push_down=true

File: be/src/exec/arrow/parquet_reader.cpp [hunk 7]
```diff
@@ -187,19 +182,25 @@ Status ParquetReaderWrap::read_record_batch(bool* eof) {
...
 
 Status ParquetReaderWrap::next_batch(std::shared_ptr<arrow::RecordBatch>* batch, bool* eof) {
-    if (_batch->num_rows() == 0 || _current_line_of_batch != 0 || _current_line_of_group != 0) {
-        RETURN_IF_ERROR(read_record_batch(eof));
+    std::unique_lock<std::mutex> lock(_mtx);
+    while (!_closed && _queue.empty()) {
+        if (_batch_eof) {
+            _include_column_ids.clear();
+            *eof = true;
+            _batch_eof = false;
```
Comment (morningman) [inline be/src/exec/arrow/parquet_reader.cpp:190]: Why set it back to false?

File: be/src/exec/arrow/parquet_reader.cpp [hunk 8]
```diff
@@ -118,6 +112,7 @@ Status ParquetReaderWrap::init_reader(const std::vector<SlotDescriptor*>& tuple_
 }
 
 void ParquetReaderWrap::close() {
+    LOG(INFO) << "ParquetReaderWrap _closed: " << _closed;
```
Comment (morningman) [inline be/src/exec/arrow/parquet_reader.cpp:115]: Change it to VLOG, or just remove it.

File: be/src/exec/arrow/parquet_reader.cpp [hunk 9]
```diff
@@ -239,8 +240,32 @@ Status ParquetReaderWrap::handle_timestamp(const std::shared_ptr<arrow::Timestam
     return Status::OK();
 }
 
+Status ParquetReaderWrap::init_parquet_type() {
```
Comment (morningman) [inline be/src/exec/arrow/parquet_reader.cpp:243]: Why extract this method out of the `init_reader()`?

File: be/src/exec/arrow/parquet_reader.cpp [hunk 10]
```diff
@@ -239,8 +240,32 @@ Status ParquetReaderWrap::handle_timestamp(const std::shared_ptr<arrow::Timestam
...
+            return Status::InternalError(_status.ToString());
+        }
+        _parquet_column_type.emplace_back(field->type()->id());
+    }
+    return Status::OK();
+}
+
 Status ParquetReaderWrap::read(Tuple* tuple, const std::vector<SlotDescriptor*>& tuple_slot_descs,
                                MemPool* mem_pool, bool* eof) {
+    if (_batch == nullptr) {
```
Comment (morningman) [inline be/src/exec/arrow/parquet_reader.cpp:265]: Remove this. I think we don't need to modify row-based engine's logic.

File: be/src/exec/arrow/parquet_reader.cpp [hunk 11]
```diff
@@ -554,6 +589,9 @@ void ParquetReaderWrap::prefetch_batch() {
 Status ParquetReaderWrap::read_next_batch() {
     std::unique_lock<std::mutex> lock(_mtx);
     while (!_closed && _queue.empty()) {
```
Comment (morningman) [inline be/src/exec/arrow/parquet_reader.cpp:591]: `read_next_batch()` is for row-based engine, no need to modify it.

File: be/src/exec/arrow/parquet_row_group_reader.cpp [hunk 12]
```diff
@@ -0,0 +1,567 @@
...
+            if (need_filter) {
+                filtered_num_row_groups++;
+                filtered_num_rows += row_group_meta->num_rows();
+                filtered_total_byte_size += row_group_meta->total_byte_size();
+                row_group_meta->schema()->name();
+                LOG(INFO) << "Filter row group id: " << row_group_id;
```
Comment (morningman) [inline be/src/exec/arrow/parquet_row_group_reader.cpp:139]: VLOG

File: be/src/exec/arrow/parquet_row_group_reader.cpp [hunk 13]
```diff
@@ -0,0 +1,567 @@
...
+            // Min-max of statistic is plain-encoded value
+            const std::string& min = statistic->EncodeMin();
+            const std::string& max = statistic->EncodeMax();
+
+            need_filter = _determine_filter_row_group(slot_iter->second, min, max);
+            if (need_filter) {
+                filtered_num_row_groups++;
+                filtered_num_rows += row_group_meta->num_rows();
+                filtered_total_byte_size += row_group_meta->total_byte_size();
+                row_group_meta->schema()->name();
```
Comment (morningman) [inline be/src/exec/arrow/parquet_row_group_reader.cpp:138]: unused row?

File: be/src/exec/arrow/parquet_row_group_reader.cpp [hunk 14]
```diff
@@ -0,0 +1,567 @@
...
+                LOG(INFO) << "Filter row group id: " << row_group_id;
+                _filter_group.emplace(row_group_id);
+            }
+        }
+    }
+    if (need_filter) {
+        _parent->statistics()->filtered_row_groups = filtered_num_row_groups;
+        _parent->statistics()->filtered_rows = filtered_num_rows;
+        _parent->statistics()->filtered_total_bytes = filtered_total_byte_size;
+        LOG(INFO) << "Parquet file: " << _file_metadata->schema()->name()
```
Comment (morningman) [inline be/src/exec/arrow/parquet_row_group_reader.cpp:148]: VLOG

File: be/src/exec/arrow/parquet_row_group_reader.cpp [hunk 15]
```diff
@@ -0,0 +1,567 @@
...
+            const std::string& max = statistic->EncodeMax();
+
+            need_filter = _determine_filter_row_group(slot_iter->second, min, max);
+            if (need_filter) {
+                filtered_num_row_groups++;
+                filtered_num_rows += row_group_meta->num_rows();
+                filtered_total_byte_size += row_group_meta->total_byte_size();
+                row_group_meta->schema()->name();
+                LOG(INFO) << "Filter row group id: " << row_group_id;
+                _filter_group.emplace(row_group_id);
```
Comment (morningman) [inline be/src/exec/arrow/parquet_row_group_reader.cpp:140]: Once this group has been filtered, we can skip the rest of conjunct and go to next row group.

File: be/src/exec/base_scanner.cpp [hunk 16]
```diff
@@ -95,6 +95,11 @@ Status BaseScanner::open() {
     return Status::OK();
 }
 
+void BaseScanner::reg_conjunct_ctxs(TupleId tupleId, std::vector<ExprContext*> conjunct_ctxs) {
```
Comment (morningman) [inline be/src/exec/base_scanner.cpp:98]:
```suggestion
void BaseScanner::reg_conjunct_ctxs(const TupleId& tupleId, const std::vector<ExprContext*>& conjunct_ctxs) {
```

File: be/src/exec/base_scanner.h [hunk 17]
```diff
@@ -142,6 +145,9 @@ class BaseScanner {
     vectorized::Block _src_block;
     int _num_of_columns_from_file;
 
+    TupleId _tupleId;
```
Comment (morningman) [inline be/src/exec/base_scanner.h:148]: Add comment to explain what this tuple id is.

File: be/src/vec/exec/file_scanner.cpp [hunk 18]
```diff
@@ -69,6 +69,11 @@ Status FileScanner::open() {
     return Status::OK();
 }
 
+void FileScanner::reg_conjunct_ctxs(TupleId tupleId, std::vector<ExprContext*> conjunct_ctxs) {
```
Comment (morningman) [inline be/src/vec/exec/file_scanner.cpp:72]:
```suggestion
void FileScanner::reg_conjunct_ctxs(const TupleId& tupleId, const std::vector<ExprContext*>& conjunct_ctxs) {
```

File: be/src/vec/exec/file_scanner.h [hunk 19]
```diff
@@ -85,6 +87,10 @@ class FileScanner {
     std::unique_ptr<vectorized::VExprContext*> _vpre_filter_ctx_ptr;
     int _num_of_columns_from_file;
 
+    // File formats based push down predicate
+    std::vector<ExprContext*> _conjunct_ctxs;
+    TupleId _tupleId;
```
Comment (morningman) [inline be/src/vec/exec/file_scanner.h:92]: Add comment for this tuple id

**Review by morningman — DISMISSED** [review]
LGTM

**Review by morningman — DISMISSED** [review]
LGTM

**Review by morningman — APPROVED** [review]

**Pull request merged**